File: front-end/Mobile/src/models/identity/Person.js
import BaseModel from '../BaseModel';

export default class Person extends BaseModel {
  constructor(data = {}) {
    super(data);
    this.personId = data.person_id || null;
    this.documentNumber = data.document_number || '';
    this.name = data.name || '';
    this.lastName = data.last_name || '';
    this.email = data.email || null;
    this.phone = data.phone || null;
    this.status = data.status !== undefined ? data.status : true;
  }

  get fullName() {
    return `${this.name} ${this.lastName}`.trim();
  }

  static fromApi(data) {
    if (!data) return null;
    return new Person(data);
  }

  toApi() {
    return {
      ...super.toApi(),
      person_id: this.personId,
      document_number: this.documentNumber,
      name: this.name,
      last_name: this.lastName,
      email: this.email,
      phone: this.phone,
      status: this.status,
    };
  }
}

File: back-end/01-ms-identity/src/main/java/com/faceattend_edu/identity_service/domain/model/Person.java
package com.faceattend_edu.identity_service.domain.model;

import lombok.AllArgsConstructor;
import lombok.Getter;
import lombok.NoArgsConstructor;
import lombok.Setter;

import java.time.LocalDate;
import java.time.LocalDateTime;
import java.util.UUID;
import java.util.regex.Pattern;

@Getter
@Setter
@AllArgsConstructor
@NoArgsConstructor
public class Person {
    private UUID personId;
    private String documentNumber;
    private String name;
    private String lastName;
    private String email;
    private String phone;
    private String documentType;
    private BloodType bloodType;
    private LocalDate birthDate;
    private String address;
    private Boolean status;
    private LocalDateTime createdAt;
    private LocalDateTime updatedAt;
    private LocalDateTime deletedAt;
    private UUID createdBy;
    private UUID updatedBy;
    private UUID deletedBy;
    private long rowVersion;

    private static final Pattern EMAIL_PATTERN = Pattern.compile("^[^@\\s]+@[^@\\s]+\\.[^@\\s]+$");

    public void validate() {
        if (isNullOrBlank(this.documentNumber)) throw new IllegalArgumentException("Person.documentNumber is required");
        if (isNullOrBlank(this.name)) throw new IllegalArgumentException("Person.name is required");
        if (isNullOrBlank(this.lastName)) throw new IllegalArgumentException("Person.lastName is required");
        if (this.email != null && !EMAIL_PATTERN.matcher(this.email).matches()) {
            throw new IllegalArgumentException("Person.email is not a valid email");
        }
        if (this.documentNumber != null && this.documentNumber.length() > 50) {
            throw new IllegalArgumentException("Person.documentNumber max 50");
        }
    }

    public String getFullName() {
        String n = this.name == null ? "" : this.name.trim();
        String ln = this.lastName == null ? "" : this.lastName.trim();
        return (n + " " + ln).trim();
    }

    public boolean isActive() {
        return Boolean.TRUE.equals(this.status);
    }

    public void touchCreatedIfMissing() {
        if (this.createdAt == null) this.createdAt = LocalDateTime.now();
    }

    private static boolean isNullOrBlank(String s) {
        return s == null || s.trim().isEmpty();
    }
}
